In [2]:
#Imnporting libraries
import pandas as pd
import numpy as np
import os
from sklearn.metrics.pairwise import cosine_similarity
from scipy.spatial.distance import cosine
from sklearn.metrics.pairwise import pairwise_distances

#Listing the existing files in the dir
for root, dirs, files in os.walk("."):  
    for filename in files:
        print(filename)

1st_assgnmt_collab.ipynb
1st_assignment_intro.ipynb
2nd_assignment_collab.ipynb
2nd_assignment_intro.ipynb
README.md
sth (2).csv
sth.csv
COMMIT_EDITMSG
config
description
FETCH_HEAD
HEAD
index
ORIG_HEAD
applypatch-msg.sample
commit-msg.sample
fsmonitor-watchman.sample
post-update.sample
pre-applypatch.sample
pre-commit.sample
pre-push.sample
pre-rebase.sample
pre-receive.sample
prepare-commit-msg.sample
update.sample
exclude
HEAD
master
master
5889fd52bdb928eb7b8ef26232706c784a7fb1
8a81d4f821e4a1d51a8c68599d74959a23bb7d
ff03ce761e5a44ccc2b2a825bf574c1bee0631
44966c920b4d862ad58a805a0ece2382922184
96b16af7d370ec7985b75b69d3af36a913954d
2974eb78edd82b431c3690f7464a503c694356
2041e1411e121aea1a74646ec651e7c9c954db
fdc207848247f45b640d1b9e61f889fd73a974
9dd7992c843ab1040b900da7e3335ed0ff5593
4eb72f8826545861a52a4848f147bf97669990
2edfd2574cdc6b5ee4d4d59f8f97f67da4b8a1
b603904c021c6626dab814b117fa373b6d11b0
498597af4392c58b0f8601759186248a25212f
a0773950660740e2fb0a2b998f594d0e94d0a1
825dc6

In [3]:
items = pd.read_csv('./files/ratings_2nd.csv')
items = items.set_index('User')
items_set = items.iloc[:20,:20]
calc_1 = items_set.multiply(items_set.iloc[:,0],axis=0).sum()
items_dist = items.iloc[20]
calc_2 = items_dist.multiply(items_dist.iloc[0], axis=0)
calc_3 = calc_1 / calc_2; calc_3.drop('1: Toy Story (1995)').sort_values(ascending=False)

260: Star Wars: Episode IV - A New Hope (1977)             0.747409
780: Independence Day (ID4) (1996)                         0.690665
296: Pulp Fiction (1994)                                   0.667846
318: Shawshank Redemption, The (1994)                      0.667424
1265: Groundhog Day (1993)                                 0.661016
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.644995
2916: Total Recall (1990)                                  0.623056
34: Babe (1995)                                            0.618070
3578: Gladiator (2000)                                     0.587852
356: Forrest Gump (1994)                                   0.580540
593: Silence of the Lambs, The (1991)                      0.570229
2028: Saving Private Ryan (1998)                           0.534579
2571: Matrix, The (1999)                                   0.505010
1259: Stand by Me (1986)                                   0.492659
1198: Raiders of the Lost Ark (1981)            

Task: Provide 5 movies most similar to Toy Story, in decreasing order of similarity using cosine similarity between items using raw (unnormalized) ratings. Only provide the movie ID, not the title.

Solution: 260, 780, 296 318, and 1265 are the most similart to Toy Story.

In [248]:
filled = items_set.fillna(0)
#Initiating the empty df
movies_sim = pd.DataFrame(index=filled.columns, columns=filled.columns)
#Calculating cosine (another apprapproach to the above)
num_titles = len(movies_sim.columns)
for i in range(0, num_titles):
    for j in range(0, num_titles):
        movies_sim.iloc[i,j] = 1 - cosine(filled.iloc[:,i], filled.iloc[:,j])  
        
mask = items_set.notnull()
transfered = items_set.mask(mask, movies_sim,axis=1)
items_set.multiply(transfered, axis=1)

#Figuring out the algorithm on the example of one movie as we know the prediction for user 755
test = items_set.loc['755']
movies_sim.iloc[0]
nom = test.multiply(movies_sim.loc['780: Independence Day (ID4) (1996)']).sum()
mask = test.notnull()
denom = test.mask(mask, movies_sim.loc['780: Independence Day (ID4) (1996)']).sum()
nom/denom

#Doing the actual calc
target = items_set.loc['5277']
merged = pd.DataFrame(target).join(movies_sim)
nominator = merged.iloc[:,1:].multiply(merged.iloc[:,0], axis=0).sum()
#Replacing those movies that the user has not rated with NaN
lst = merged.iloc[:,0][merged.iloc[:,0].isnull()].reset_index().iloc[:,0].tolist()
for col in lst:
    merged.loc[:,[col]] = np.nan
denominator = merged.iloc[:,1:].sum(axis=1)
calc = nominator / denominator; calc.sort_values(ascending=False).head(5)
#Calculating user and item average
#target_user_mean = items_set.mean(axis=1).loc['5277']
#user_mean = items_set.mean(axis=1)
#items_mean = items_set.mean(axis=0)

527: Schindler's List (1993)                      2.973883
1259: Stand by Me (1986)                          2.928801
260: Star Wars: Episode IV - A New Hope (1977)    2.922240
593: Silence of the Lambs, The (1991)             2.883304
2396: Shakespeare in Love (1998)                  2.852131
dtype: float64

Task: Provide the top 5 recommended movies for user 5277, using an average of the user’s ratings weighted by similarity to each candidate movie. You do not need to exclude movies they have rated. Consider all movies with nonnegative similarities (do not limit neighborhood size). Similarity between items is estimated using cosine similarity between items using raw (unnormalized) ratings.

Solution: The top five movies for user 5277 are 527, 1259, 260, 593 and 2396.

In [306]:
#Normalizing scores by the item average
norm_rating = items_set.subtract(items_mean, axis=1).fillna(0)
calc_1 = norm_rating.multiply(norm_rating.iloc[:,0],axis=0).sum()
eql_dist = np.sqrt(np.sum(np.square(norm_rating)))
calc_2 = eql_dist.multiply(eql_dist.iloc[0], axis=0)
calc_3 = calc_1 / calc_2; calc_3.drop('1: Toy Story (1995)').sort_values(ascending=False).head(5)

34: Babe (1995)                          0.540738
318: Shawshank Redemption, The (1994)    0.387806
296: Pulp Fiction (1994)                 0.383212
2028: Saving Private Ryan (1998)         0.282144
356: Forrest Gump (1994)                 0.272193
dtype: float64

Task: Provide 5 movies most similar to Toy Story, in decreasing order of similarity, by computing item similarities using cosine similarity using adjusted (normalized) ratings. Only provide the movie ID, not the title.

Solution: 34, 318, 296, 2028, and 356 are the most similart to Toy Story.

Task: Provide the top 5 recommended movies for user 5277, using an average of the user’s ratings weighted by similarity to each candidate movie. You do not need to exclude movies they have rated. Consider all movies with nonnegative similarities (do not limit neighborhood size). Similarity between items is estimated using cosine similarity between items using adjusted (normalized) ratings.

Solution: